src/gwt/src/org/rstudio/studio/client/workbench/views/tutorial/TutorialPane.java

```diff
@@ -409,6 +447,74 @@ private final native void initTutorialJsCallbacks()
       
    }-*/;
    
+   private final native void initExternalWindowJsCallbacks(WindowEx window,
+                                                           String tutorialUrl,
+                                                           String tutorialName,
+                                                           String tutorialPackage)
+   /*-{
+      
+      // register this window
+      $wnd.tutorialWindows = $wnd.tutorialWindows || {};
+      $wnd.tutorialWindows[tutorialUrl] = {
+         "package": tutorialPackage,
+         "name": tutorialName,
+         "window": window
+      };
+      
+      // start polling for window closure
+      var self = this;
+      $wnd.tutorialWindowsCallback = $wnd.tutorialWindowsCallback || setInterval(function() {
+         
+         // stop any tutorials whose associated window was closed
+         for (var url in $wnd.tutorialWindows)
+         {
+            var entry = $wnd.tutorialWindows[url];
+            
+            var window = entry["window"];
+            if (window.closed)
+            {
+               self.@org.rstudio.studio.client.workbench.views.tutorial.TutorialPane::stopTutorial(*)(url);
+               delete $wnd.tutorialWindows[url];
+            }
+         }
+         
+         // stop polling if we have no more child windows
+         var keys = Object.keys($wnd.tutorialWindows);
+         if (keys.length === 0)
+         {
+            clearInterval($wnd.tutorialWindowsCallback);
+            $wnd.tutorialWindowsCallback = null;
+         }
+         
+      }, 500);
```

Comment:
This is fine, but could be made more responsive (and avoid polling) by having the tutorial window notify us when it's closed. Is that feasible here? (IIRC we usually do this with something like ShinyDisconnectNotifier, but not sure if that's practical here since we don't wrap the tutorial in a frame we control)

Reply:
Right -- we could just serve a separate page that wraps the tutorial in an `<iframe>` rather than navigating to the application directly, but that feels more expensive than I'd want to do for this PR.